test/gas/gas.js

```diff
@@ -0,0 +1,563 @@
+
+        const { gasUsed, cumulativeGasUsed } = await method.send({ from, gas: gasEstimate });
+
+        if (provider.options.hardfork === "byzantium") {
+          // since we are resetting to a non-zero value, there is no gas added to the refund counter here
+          assert.strictEqual(gasUsed, gasEstimate);
+        } else if (provider.options.hardfork === "constantinople") {
+          // since storage was initially primed to 1 and we call triggerRsclearRefundForY(), which then
+          // resets storage back to 1, 4800 gas is added to the refund counter per Constantinople EIP 1283
+          assert.strictEqual(gasUsed, gasEstimate - rsclearRefundForResettingDirtySlotToNonZeroValue);
+        }
+        assert.strictEqual(gasUsed, cumulativeGasUsed);
+      }
+    );
+
+    it(
+      "accounts for Rsclear Refund in gasEstimate when a fresh storage slot's original " +
+        "value is not 0 and new value is 0",
+      async() => {
+        const { accounts, instance, provider } = services;
+        const from = accounts[0];
+        const rsclearRefundForUpdatingFreshSlotToZero = 15000;
+        const options = { from, gas: 5000000 };
+
+        // prime storage by making sure storage is set to 1
+        await instance.methods.initialSettingOfX().send(options);
+
+        // update storage to be 0
+        const method = instance.methods.reset();
+
+        const gasEstimate = await method.estimateGas(options);
+
+        const { gasUsed, cumulativeGasUsed } = await method.send({ from, gas: gasEstimate });
+
+        if (provider.options.hardfork === "byzantium") {
+          assert.strictEqual(gasUsed, gasEstimate - RSCLEAR_REFUND);
+        } else if (provider.options.hardfork === "constantinople") {
+          // since storage was initially primed to 1 and we call reset(), which then sets
+          // storage to 0, 15000 gas is added to the refund counter per Constantinople EIP 1283
+          assert.strictEqual(gasUsed, gasEstimate - rsclearRefundForUpdatingFreshSlotToZero);
+        }
+        assert.strictEqual(gasUsed, cumulativeGasUsed);
+      }
+    );
+
+    it(
+      "accounts for Rsclear Refund in gasEstimate when a dirty storage slot's original value " +
+        "is not 0 and new value is 0",
+      async() => {
+        const { accounts, instance, provider } = services;
+        const from = accounts[0];
+        const rsclearRefundForUpdatingDirtySlotToZero = 15000;
+        const options = { from, gas: 5000000 };
+
+        // prime storage by making sure storage is set to 1
+        await instance.methods.initialSettingOfX().send(options);
+
+        // update storage and then reset it to 0
+        const method = instance.methods.triggerRsclearRefund();
+
+        const gasEstimate = await method.estimateGas(options);
+
+        const { gasUsed, cumulativeGasUsed } = await method.send({ from, gas: gasEstimate });
+
+        if (provider.options.hardfork === "byzantium") {
```

Comment:
do we test both fork types?

Reply:
Currently only testing `constantinople`.

Reply:
Added byzantium support to this test file.